README.md
# Easyship Node.js API wrapper
[![npm version](https://badge.fury.io/js/easyship.svg)](https://badge.fury.io/js/easyship)
[![codecov](https://codecov.io/gh/dplewis/easyship-node/branch/master/graph/badge.svg?token=2pNFBkoOnu)](https://codecov.io/gh/dplewis/easyship-node?branch=master)
[![Build Status](https://travis-ci.org/dplewis/easyship-node.svg?branch=master)](https://travis-ci.org/dplewis/easyship-node)

You will need to [register for a Easyship account][easyship] to use the Easyship API. It's free to sign up, free to use the API. Only pay to print a live label, test labels are free.

## Installation:
You can install this package by running the following command:
```shell
  npm install easyship
```

## Usage:

Initialize your `easyship` instance using your `API Access Token` provided to you on the `API` page in the Easyship Dashboard.

```js
  const easyship = require('easyship')('<ACCESS_TOKEN>');

  easyship.rate.create({
    origin_postal_code: '60605',
    destination_country_alpha2: 'US',
    destination_postal_code: '60605',
    taxes_duties_paid_by: 'Sender',
    is_insured: false,
    apply_shipping_rules: true,
    items: [{
      actual_weight: 1.2,
      height: 10,
      width: 15,
      length: 20,
      category: 'mobiles',
      declared_currency: 'SGD',
      declared_customs_value: 100,
    }],
  }).then(({ rates }) => {
    console.log(rates);
  });
```

## Method Overview

## [Shipping API][api-reference]

HTTP Method | Endpoint                                      | Methods
----------- | --------------------------------------------- | ----------------------------
POST        | /rate/v1/rates                                | easyship.rate.create()
POST        | /shipment/v1/shipments                        | easyship.shipment.create()
POST        | /shipment/v1/shipments/create_and_buy_label   | easyship.shipment.createLabel()
PATCH       | /shipment/v1/shipments/{easyship_shipment_id} | easyship.shipment.update()
GET         | /shipment/v1/shipments/{easyship_shipment_id} | easyship.shipment.retrieve()
GET         | /shipment/v1/shipments                        | easyship.shipment.list()
DELETE      | /shipment/v1/shipments/{easyship_shipment_id} | easyship.shipment.del()
PATCH       | /shipment/v1/shipments/update_warehouse_state | easyship.shipment.updateWarehouse()
POST        | /label/v1/labels                              | easyship.label.create()
GET         | /pickup/v1/pickup_slots/{courier_id}          | easyship.pickup.retrieve()
POST        | /pickup/v1/pickups                            | easyship.pickup.create()
POST        | /pickup/v1/direct_handover                    | easyship.pickup.directHandover()
GET         | /track/v1/status                              | easyship.track.status()
GET         | /track/v1/checkpoints                         | easyship.track.checkpoints()
GET         | /reference/v1/categories                      | easyship.categories.list()

Check [example.js][example] for more detail.

## Credits

This project was created from [stripe-node][stripe].

## Documentation

Please see [https://developers.easyship.com/reference][api-reference] documentation.

## Sponsor

<a href="https://www.buymeacoffee.com/dplewis" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: 41px !important;width: 174px !important;box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;-webkit-box-shadow: 0px 3px 2px 0px rgba(190, 190, 190, 0.5) !important;" ></a>

[api-reference]: https://developers.easyship.com/reference
[easyship]: https://easyship.com/
[example]: https://github.com/dplewis/easyship-node/blob/master/example.js
[stripe]: https://github.com/stripe/stripe-node


CHANGELOG.md
### master
[Full Changelog](https://github.com/dplewis/easyship-node/compare/1.0.0...master)

### 1.0.0 (10/27/2020)

* Initial Release


example.js
//  const easyship = require('easyship')('<API KEY>');
const easyship = require('./lib/easyship.js')('');

// POST /rate/v1/rates
easyship.rate.create({
  origin_postal_code: '60605',
  destination_country_alpha2: 'US',
  destination_postal_code: '60605',
  taxes_duties_paid_by: 'Sender',
  is_insured: false,
  apply_shipping_rules: true,
  items: [{
    actual_weight: 1.2,
    height: 10,
    width: 15,
    length: 20,
    category: 'mobiles',
    declared_currency: 'SGD',
    declared_customs_value: 100,
  }],
}).then(({ rates }) => {
  console.log(rates);
}).catch((e) => {
  console.log(e);
});

// POST /shipment/v1/shipments
// easyship.shipment.create({
//   platform_name: 'Amazon',
//   platform_order_number: '#1234',
//   selected_courier_id: 'b6f1d8c1-7e1e-465e-aa64-880d34662e54',
//   destination_country_alpha2: 'US',
//   destination_city: 'New York',
//   destination_postal_code: '10022',
//   destination_state: 'NY',
//   destination_name: 'Aloha Chen',
//   destination_company_name: 'My Company',
//   destination_address_line_1: '300 Park Avenue',
//   destination_address_line_2: null,
//   destination_phone_number: '+1 234-567-890',
//   destination_email_address: 'api-support@easyship.com',
//   items: [{
//     description: 'Silk dress',
//     sku: 'test',
//     actual_weight: 1.2,
//     height: 10,
//     width: 15,
//     length: 20,
//     category: 'mobiles',
//     declared_currency: 'SGD',
//     declared_customs_value: 100,
//   }],
// }).then(({ shipment }) => {
//   console.log(shipment);
//   console.log(shipment.easyship_shipment_id);
// }).catch((e) => {
//   console.log(e);
// });

// POST /shipment/v1/shipments/create_and_buy_label
// easyship.shipment.createLabel({
//   platform_name: 'Amazon',
//   platform_order_number: '#1234',
//   selected_courier_id: 'b6f1d8c1-7e1e-465e-aa64-880d34662e54',
//   destination_country_alpha2: 'US',
//   destination_city: 'New York',
//   destination_postal_code: '10022',
//   destination_state: 'NY',
//   destination_name: 'Aloha Chen',
//   destination_company_name: 'My Company',
//   destination_address_line_1: '300 Park Avenue',
//   destination_address_line_2: null,
//   destination_phone_number: '+1 234-567-890',
//   destination_email_address: 'api-support@easyship.com',
//   items: [{
//     description: 'Silk dress',
//     sku: 'test',
//     actual_weight: 1.2,
//     height: 10,
//     width: 15,
//     length: 20,
//     category: 'mobiles',
//     declared_currency: 'SGD',
//     declared_customs_value: 100,
//   }],
// }).then(({ shipment }) => {
//   console.log(shipment);
//   console.log(shipment.easyship_shipment_id);
// }).catch((e) => {
//   console.log(e);
// });

// PATCH /shipment/v1/shipments/{easyship_shipment_id}
// easyship.shipment.update('ESUS10043353', {
//   destination_name: 'Node',
// }).then(({ shipment }) => {
//   console.log(shipment);
// }).catch((e) => {
//   console.log(e);
// });

// GET /shipment/v1/shipments/{easyship_shipment_id}
// easyship.shipment.retrieve('ESUS10043348').then(({ shipment }) => {
//   console.log('shipment');
//   console.log(shipment);
// }).catch((e) => {
//   console.log(e);
// });

// GET /shipment/v1/shipments
// easyship.shipment.list({
//   easyship_shipment_id: 'ESUS10043350',
// }).then(({ shipments, total_pages, current_page }) => {
//   console.log(total_pages, current_page, shipments);
// }).catch((e) => {
//   console.log(e);
// });

// DELETE /shipment/v1/shipments/{easyship_shipment_id}
// easyship.shipment.del('ESUS10043351').then((response) => {
//   console.log(response.message);
// }).catch((e) => {
//   console.log(e);
// });

// PATCH /shipment/v1/shipments/update_warehouse_state
// easyship.shipment.updateWarehouse({
//   easyship_shipment_ids: ['ESUS10043354'],
//   warehouse_state: 'packed',
// }).then((response) => {
//   console.log(response.message);
// }).catch((e) => {
//   console.log(e);
// });

// POST /label/v1/labels
// easyship.label.create({
//   shipments: [{ easyship_shipment_id: 'ESUS10043358' }],
// }).then((response) => {
//   console.log(response);
// }).catch((e) => {
//   console.log(e);
// });

// GET /pickup/v1/pickup_slots/{courier_id}
// easyship.pickup.retrieve('b6f1d8c1-7e1e-465e-aa64-880d34662e54').then(({ pickup }) => {
//   console.log(pickup);
// }).catch((e) => {
//   console.log(e);
// });

// POST /pickup/v1/pickups
// easyship.pickup.create({
//   courier_id: 'b6f1d8c1-7e1e-465e-aa64-880d34662e54',
//   preferred_date: '2021-12-08',
//   preferred_max_time: '2021-12-08T18:00',
//   preferred_min_time: '2021-12-08T09:00',
//   easyship_shipment_ids: [
//     'ESUS10043358',
//   ],
// }).then((response) => {
//   console.log(response);
// }).catch((e) => {
//   console.log(e);
// });

// POST /pickup/v1/pickups
// easyship.pickup.directHandover({
//   easyship_shipment_ids: [
//     'ESUS10043358',
//   ],
// }).then((response) => {
//   console.log(response);
// }).catch((e) => {
//   console.log(e);
// });

// GET /track/v1/status
// easyship.track.status({
//   easyship_shipment_id: 'ESUS10043358',
// }).then((response) => {
//   console.log(response);
// }).catch((e) => {
//   console.log(e);
// });

// GET /track/v1/checkpoints
// easyship.track.checkpoints({
//   easyship_shipment_id: 'ESUS10043358',
// }).then((response) => {
//   console.log(response);
// }).catch((e) => {
//   console.log(e);
// });

// GET /reference/v1/categories
// easyship.categories.list().then(({ categories }) => {
//   console.log(categories);
// }).catch((e) => {
//   console.log(e);
// });
